packages/api/src/external/aws/sqs.ts

```diff
@@ -62,5 +63,189 @@ export async function sendMessageToQueue(
+}
+
+export type PeekMessageSQSParameters = SharedInternalGetMessageSQSParameters;
+
+/**
+ * Reads messages from the queue, returns them, but does not remove them.
+ * Only queries the queue once.
+ */
+export async function peekMessagesFromQueue(
+  queueUrl: string,
+  sqsParams: GetMessageSQSParameters
+): Promise<SQS.Message[]> {
+  return _getMessagesFromQueue(queueUrl, {
+    ...sqsParams,
+    removeMessages: false,
+    poolUntilEmpty: false,
+  });
+}
+
+type SharedInternalGetMessageSQSParameters = {
+  maxNumberOfMessagesPerQuery?: number;
+  visibilityTimeout?: number;
+  waitTimeSeconds?: number;
+};
+type InternalGetMessageSQSParameters = SharedInternalGetMessageSQSParameters & {
+  maxNumberOfMessages?: number;
+  removeMessages: boolean;
+  poolUntilEmpty: boolean;
+};
+
+async function _getMessagesFromQueue(
+  queueUrl: string,
+  sqsParams: InternalGetMessageSQSParameters,
+  rollingResult: SQS.Message[] = []
+): Promise<SQS.Message[]> {
+  const {
+    removeMessages,
+    poolUntilEmpty,
+    maxNumberOfMessages = 10,
+    maxNumberOfMessagesPerQuery = 1,
+  } = sqsParams;
+  if (!removeMessages && poolUntilEmpty) {
+    throw new Error("Cannot pool until empty if not removing messages");
+  }
+
+  const remainingMessagesToQuery = maxNumberOfMessages - rollingResult.length;
+  const maxMessagesOnQuery = Math.min(maxNumberOfMessagesPerQuery, remainingMessagesToQuery);
+  if (maxMessagesOnQuery <= 0) return rollingResult;
+
+  const messageParams: SQS.Types.ReceiveMessageRequest = {
+    QueueUrl: queueUrl,
+    MaxNumberOfMessages: maxMessagesOnQuery,
+    VisibilityTimeout: sqsParams.visibilityTimeout,
+    WaitTimeSeconds: sqsParams.waitTimeSeconds ?? 1,
+    AttributeNames: ["All"],
+    MessageAttributeNames: ["All"],
+  };
+
+  const resultReceive = await sqs.receiveMessage(messageParams).promise();
+  const messages = resultReceive.Messages ?? [];
+  const result = [...rollingResult, ...messages];
+
+  if (removeMessages) await deleteMessagesFromQueue(queueUrl, messagesToDelete(messages));
+
+  if (poolUntilEmpty) {
+    const updatedTotal = await getMessageCountFromQueue(queueUrl);
+    if (updatedTotal <= 0) return result;
+    return _getMessagesFromQueue(queueUrl, sqsParams, result);
+  }
+  return result;
+}
+
+/**
+ * Returns the approximate count of messages in the queue. Returns -1 if not available.
+ */
+export async function getMessageCountFromQueue(queueUrl: string): Promise<number> {
+  const resultTotal = await sqs
+    .getQueueAttributes({
+      QueueUrl: queueUrl,
+      AttributeNames: ["ApproximateNumberOfMessages"],
+    })
+    .promise();
+  const total = Number(resultTotal.Attributes?.ApproximateNumberOfMessages ?? "-1");
+  return total;
+}
+
+function messagesToDelete(messages: SQS.Message[]): MessageToDelete[] {
+  return messages.flatMap(m => {
+    if (!m.MessageId || !m.ReceiptHandle) return [];
+    return {
+      id: m.MessageId,
+      receiptHandle: m.ReceiptHandle,
+    };
+  });
+}
```

Comment:
what's this doing?

by default I'd assume we'd just delete all the messages we've read from the queue, why would there ever by one without the `messageId` or `ReceiptHandle`?

Reply:
You mean `messagesToDelete()`, right? This is a function to convert from one format another. I can rename the function, agree its not the best name. 😄